### ISSUE TITLE
Why is instanceof and getClass not allowed? What are the risks to using these or benefits to not using them?

### ISSUE THREAD

**Issue by ezeAng** [nus-cs2030/2122-s1#70]

**Comment by albertsutz** [comment]
For cs2030 alone, I believe that we are trying to not use bad programming practices (typecasting when not needed, etc). As you usually do typecasting when doing instanceof, preventing you to not use instanceof is to help you not do these bad practices.

**Comment by Enigmatrix** [comment]
As for `getClass()`, the primary use case is this is for reflection, and that is generally frowned upon in this module, I think. More info over [here](https://github.com/nus-cs2030/2122-s1/issues/56#issuecomment-922187486).

**Comment by EricBryann** [comment]
This site here gives very good answers:
https://www.quora.com/Is-using-instanceof-in-Java-consider-bad-practice-Any-alternative-to-using-this-keyword

tldr:
`instanceof` is a sign of bad code design, and most of the time can be redesigned in a better way.

**Comment by yuqitanyq** [comment]
I think it's mostly to discipline us to think about better design and in general, sort of know what we are doing when we write programs. I think it's just a good practice thing.

**Comment by chngchngchng** [comment]
I would believe that we are encouraged to make use of polymorphism and overriding to decide which methods to call / what to do to different objects at runtime, and calling`instanceof` would be a hacky way to do this even if our classes were not designed cleanly.

**Comment by lulucopter** [comment]
I think the main reason we are not encouraged to use instanceof is usually because logically this means our code are designed wrongly. Either our subclass fulfils a different purpose, which begs the question why not we just place it into another object class. I would only see possible use of it when we override equals()

**Comment by KeithPJX** [comment]
I think it is fine to use instanceof as long as it is within the class that you are checking for. I guess this exception that I've seen is due to the equals method?

**Comment by DerenC** [comment]
Say u have Class B, Class C inheriting from Class A (or alternatively implementing an Interface A).

If you allow the client to use instanceof to check for the run-time type of an object with a compile-time type of A, before the client processes it, aren't you violating **tell-dont-ask principle**. 

A better design would probably be, like what others have suggested, to have different implementation for some overriding methods in Class B & C respectively. If a client receives an A object, he is not supposed to know is it actually a B or C object, left alone processing it himself based on the runtime type he used instanceof to check.
